lib/utils/device/live.ts

```diff
@@ -179,8 +179,9 @@ export class LivepushManager {
 					stageImages: this.imageIds[serviceName],
 					docker: this.docker,
 				});
-				const buildVars = buildTask.buildMetadata.getBuildVarsForService(
-					buildTask.serviceName,
+				const buildVars = _.merge(
+					buildTask.buildMetadata.getBuildVarsForService(buildTask.serviceName),
+					buildTask.args,
```

Comment:
Here's what I am thinking. Users can already specify build variables through the `balena.yml` file, which can be used with both livepush (`push <ip-address>`) and cloud push (`push myApp`). The `getBuildVarsForService()` call above takes the build variables from the `balena.yml` file in the build tar stream. The solution currently proposed in this PR goes around that mechanism, and implements something that only works for livepush, not cloud push.

The `--buildArg` flag documentation in this PR, as it is, does not even specify that the feature only works for `balena push <ip-address>`, and not for `balena push myApp`.

We could modify the documentation of the `--buildArg` flag to clarify the restriction, but I am thinking that we could change the implementation so that it works for both local and cloud push.

I do **not** propose that the CLI should to modify any files on disk (the `./.balena/balena.yml` file). However, it could modify the tar stream on the fly like we already do for other purposes, such as converting line endings on Windows:

https://github.com/balena-io/balena-cli/blob/v12.30.4/lib/utils/compose_ts.ts#L658-L664

If we find the `.balena/balena.yml` file in the tar stream, we update it to add the build variables specified with `--buildArg` on the command line (such that the command line takes precedence over any variables of same name that might exist in the file). If we don't find `.balena/balena.yml` in the tar stream, we add a new one.

We should also accept the variations of the file name:
https://github.com/balena-io-modules/resin-multibuild/blob/v4.10.0/lib/build-metadata.ts#L92-L98
```
			{ name: 'balena.yml', type: MetadataFileType.Yaml },
			{ name: 'balena.yaml', type: MetadataFileType.Yaml },
			{ name: 'balena.json', type: MetadataFileType.Json },
			{ name: 'resin.yml', type: MetadataFileType.Yaml },
			{ name: 'resin.yaml', type: MetadataFileType.Yaml },
			{ name: 'resin.json', type: MetadataFileType.Json },
```

Editing the tar stream in this way is a bit more work than the solution currently proposed in this PR, but it is a solution that would work with both local push and cloud push, and would avoid code duplication overall.

`balena.yml` reference:
https://www.balena.io/docs/learn/deploy/deployment/#build-time-secrets-and-variables

Reply:
@pdcastro Thanks, I'll take a look at trying that out next week.

Reply:
This will be very useful and exactly what we're missing in our development process.

Reply:
@iamhosseindhv, if helpful, note that currently it is possible to specify build variables and secrets using a `.balena/balena.yml` file, as documented on this page: https://www.balena.io/docs/learn/deploy/deployment/#build-time-secrets-and-variables
The `--buildArg` flag to be added in this PR would provide an alternative, more convenient way of doing the same.